timemachine/cpp/src/verlet_integrator.cu

```diff
@@ -0,0 +1,119 @@
+
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "math_utils.cuh"
+#include "verlet_integrator.hpp"
+
+#include "kernels/k_integrator.cuh"
+
+namespace timemachine {
+
+VelocityVerletIntegrator::VelocityVerletIntegrator(int N, double dt, const double *h_cbs)
+    : N_(N), dt_(dt), initialized_(false) {
+
+    d_cbs_ = gpuErrchkCudaMallocAndCopy(h_cbs, N);
+    gpuErrchk(cudaMalloc(&d_du_dx_, N * 3 * sizeof(*d_du_dx_)));
+}
+
+VelocityVerletIntegrator::~VelocityVerletIntegrator() {
+    gpuErrchk(cudaFree(d_cbs_));
+    gpuErrchk(cudaFree(d_du_dx_));
+}
+
+void VelocityVerletIntegrator::step_fwd(
+    std::vector<BoundPotential *> &bps,
+    double lamb,
+    double *d_x_t,
+    double *d_v_t,
+    double *d_box_t,
+    unsigned long long *d_du_dl,
+    cudaStream_t stream) {
+
+    gpuErrchk(cudaMemsetAsync(d_du_dx_, 0, N_ * 3 * sizeof(*d_du_dx_), stream));
+
+    const int D = 3;
+    size_t tpb = warp_size;
+    size_t n_blocks = ceil_divide(N_, tpb);
+    dim3 dimGrid_dx(n_blocks, D);
+    for (int i = 0; i < bps.size(); i++) {
+        bps[i]->execute_device(
+            N_, d_x_t, d_box_t, lamb, d_du_dx_, nullptr, d_du_dl ? d_du_dl : nullptr, nullptr, stream);
+    }
+    update_forward_velocity_verlet<double><<<dimGrid_dx, tpb, 0, stream>>>(N_, D, d_cbs_, d_x_t, d_v_t, d_du_dx_, dt_);
+    gpuErrchk(cudaPeekAtLastError());
+}
+
+void VelocityVerletIntegrator::initialize(
+    std::vector<BoundPotential *> &bps,
+    double lamb,
+    double *d_x_t,
+    double *d_v_t,
+    double *d_box_t,
+    cudaStream_t stream) {
+
+    if (initialized_) {
+        // No op if this is initialized, will just mess up the velos
+        return;
```

Comment:
Could this actually just throw a runtime exception? Are there any cases where calling this twice before finalize isn't a bug?